synapse/appservice/scheduler.py

```diff
@@ -254,12 +246,11 @@ def _backoff(self):
             self.backoff_counter += 1
         self.recover()
 
-    @defer.inlineCallbacks
-    def retry(self):
+    async def retry(self):
         logger.info("Starting retries on %s", self.service.id)
         try:
             while True:
-                txn = yield self.store.get_oldest_unsent_txn(self.service)
+                txn = await self.store.get_oldest_unsent_txn(self.service)
                 if not txn:
                     # nothing left: we're done!
                     self.callback(self)
```

Comment:
As far as I can tell this `self.callback` is [`_TransactionController.on_recovered`](https://github.com/matrix-org/synapse/blob/72bc285669f9ccf0e95bb43afcdaf052845a3431/synapse/appservice/scheduler.py#L180-L189), which used to be a `inlineCallbacks` (is now async), but doesn't get awaited on. This seems wrong.

Note that this class is constructed [here](https://github.com/matrix-org/synapse/blob/72bc285669f9ccf0e95bb43afcdaf052845a3431/synapse/appservice/scheduler.py#L207-L209).

Reply:
Looks like this started as a normal function and then turned into a deferred: https://github.com/matrix-org/synapse/commit/f260cb72cd3435d540411962a92ca2a9fd333eb1#diff-b1805daad13332892217490e8c31cc0cL169-L171 without the call being modified: https://github.com/matrix-org/synapse/commit/f260cb72cd3435d540411962a92ca2a9fd333eb1#diff-b1805daad13332892217490e8c31cc0cL231

Reply:
I added an await here. 👍